File: JAVA/3Sum Closest.java
public class Solution {
    public int threeSumClosest(int[] num, int target) {
        Arrays.sort(num);
        int len = num.length;
        int diff = target - (num[0] + num[1] + num[2]);
        for (int i = 0; i < len - 2; i++)
        {
            int tmp = twoSumClosest(num, target - num[i], i + 1);
            if (tmp == 0) return target;
            if (Math.abs(tmp) < Math.abs(diff))
            {
                diff = tmp;
            }
        }
        return target - diff;
    }
     
    private int twoSumClosest(int[] num, int target, int l)
    {
        int r = num.length - 1;
        int diff = target - num[l] - num[r];
        while (l < r)
        {
            int sum = num[l] + num[r];
            if (sum == target)
            {
                return 0;
            }
            else
            {
                int tmp = target - num[l] - num[r];
                if (Math.abs(tmp) < Math.abs(diff))
                {
                    diff = tmp;
                }
                if (sum > target)
                {
                    r--;
                }
                else
                {
                    l++;
                }
            }
        }
        return diff;
    }
}

File: 3Sum Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int> &num, int target) {
        sort(num.begin(), num.end());
        int size = num.size();
        int diff = INT_MAX;
        int result = 0;
        for (int i = 0; i < size - 2; i++) {
            int j = i + 1;
            int k = size - 1;
            int sum = target - num[i];
            while (j < k) {
                int tmp = num[j] + num[k];
                if (tmp == sum) {
                    return target;
                } else {
                    int tmp_diff = abs(tmp - sum);
                    if (tmp_diff < diff) {
                        result = num[i] + tmp;
                        diff = tmp_diff;
                    }
                    if (tmp < sum) {
                        j++;
                    } else {
                        k--;
                    }
                }
            }
        }
        return result;
    }
};
